src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1195,16 +1195,155 @@ private ImplicitNamedTypeSymbol BindScriptClass()
+                {
+                    var previousTarget = previousUsing.NamespaceOrType;
+                    if (previousTarget.IsType)
+                    {
+                        if (uniqueTargets.Add(previousTarget))
+                        {
+                            externalUsings.Add(previousUsing);
+                        }
+                    }
+                    else
+                    {
+                        var expandedNamespace = ExpandPreviousSubmissionNamespace((NamespaceSymbol)previousTarget, this.GlobalNamespace);
+                        if (uniqueTargets.Add(expandedNamespace))
+                        {
+                            externalUsings.Add(new NamespaceOrTypeAndUsingDirective(expandedNamespace, previousUsing.UsingDirective));
+                        }
+                    }
+                }
+            }
+
+            foreach (var newUsing in newUsings)
+            {
+                if (uniqueTargets.Add(newUsing.NamespaceOrType))
+                {
+                    externalUsings.Add(newUsing);
+                }
+            }
+
+            uniqueTargets.Free();
+
+            return externalUsings.ToImmutableAndFree();
+        }
+
+        private static NamespaceSymbol ExpandPreviousSubmissionNamespace(NamespaceSymbol originalNamespace, NamespaceSymbol expandedGlobalNamespace)
+        {
+            // Soft assert: we'll still do the right thing if it fails.
+            Debug.Assert(!originalNamespace.IsGlobalNamespace, "Global using to global namespace");
+
+            var nameParts = ArrayBuilder<string>.GetInstance();
+            var curr = originalNamespace;
+            while (!curr.IsGlobalNamespace)
+            {
+                nameParts.Add(curr.Name);
+                curr = curr.ContainingNamespace;
+            }
+
+            var expandedNamespace = expandedGlobalNamespace;
+            for (int i = nameParts.Count - 1; i >= 0; i--)
+            {
+                // Note, the name may have become ambiguous (e.g. if a type with the same name
+                // is now in scope), but we're not rebinding - we're just expanding to the
+                // current contents of the same namespace.
+                expandedNamespace = expandedNamespace.GetMembers(nameParts[i]).OfType<NamespaceSymbol>().Single();
+            }
+            nameParts.Free();
+
+            return expandedNamespace;
         }
 
-        internal IEnumerable<NamespaceOrTypeSymbol> GlobalUsings
+        /// <summary>
+        /// Using aliases declared by this submission (null if this isn't one).
+        /// </summary>
+        /// <remarks>
+        /// Always consider whether usings should be considered before accessing this (e.g. not when binding other usings).
+        /// </remarks>
+        internal Dictionary<string, AliasAndUsingDirective> SubmissionUsingAliases => SubmissionImports?.UsingAliases;
```

Comment:
Could we use a non-null ImmutableDictionary for Imports.UsingAliases instead, to avoid the need for null checks, and avoid the perception that the collection is mutable?